server/src/main/java/org/elasticsearch/transport/ProxyConnectionStrategy.java

```diff
@@ -260,7 +260,12 @@ public void onResponse(Void v) {
                 @Override
                 public void onFailure(Exception e) {
                     if (countDown.countDown()) {
-                        openConnections(finished, attemptNumber + 1);
+                        if (attemptNumber >= MAX_CONNECT_ATTEMPTS_PER_RUN && connectionManager.size() == 0) {
+                            logger.warn(() -> "failed to open any proxy connections to cluster [" + clusterAlias + "]", e);
+                            finished.onFailure(new NoSeedNodeLeftException(strategyType(), clusterAlias, e));
+                        } else {
+                            openConnections(finished, attemptNumber + 1);
```

Comment:
This change basically tries to mirror how [error is handled in the sniff mode](https://github.com/elastic/elasticsearch/blob/696e63570fadbff061c2530bcdcd62306e542e50/server/src/main/java/org/elasticsearch/transport/SniffConnectionStrategy.java#L232-L245) so that the underlying cause of the failure is retained and reported.

Reply:
With this change, do we still need to check `connectionManager.size()` in the `else` branch below? I.e. can we remove this cause-free exception?

```
finished.onFailure(new NoSeedNodeLeftException(strategyType(), clusterAlias));
```

Or maybe we should have the `attemptNumber >= MAX_CONNECT_ATTEMPTS_PER_RUN` check in the `onResponse` handler above too.

Reply:
> With this change, do we still need to check connectionManager.size() in the else branch below? 

In practice, we should not need it. But in theory, it is possible that the `if (attemptNumber <= MAX_CONNECT_ATTEMPTS_PER_RUN)` check fails in the first invocation and goes to the `else` branch.

We have a similar [else](https://github.com/elastic/elasticsearch/blob/2a1da6d92163b2bd7b69a40a74a6736237800ae3/server/src/main/java/org/elasticsearch/transport/SniffConnectionStrategy.java#L349-L350) branch in SniffConnectionStrategy as well. Therefore I didn't remove it.

Reply:
Right, yes, but since this only affects the first invocation (and only if there are no seed nodes or some other weird config?) I think it'd be better to handle it up-front. That would clarify that once we get here we never return a cause-free `NoSeedNodeLeftException`.